objects/src/accounts/code.rs

```diff
@@ -123,36 +129,44 @@ impl AccountCode {
     /// Returns the procedure index for the procedure with the specified root or None if such
     /// procedure is not defined for this account.
     pub fn get_procedure_index_by_root(&self, root: Digest) -> Option<usize> {
-        let root_bytes = root.as_bytes();
-        self.procedures.binary_search_by(|x| x.as_bytes().cmp(&root_bytes)).ok()
+        self.procedures.iter().position(|r| r == &root)
     }
 }
 
-// SERIALIZATION
+// EQUALITY
 // ================================================================================================
 
-#[cfg(feature = "serde")]
-mod serialization {
-    use assembly::ast::{AstSerdeOptions, ModuleAst};
+impl PartialEq for AccountCode {
+    fn eq(&self, other: &Self) -> bool {
+        // TODO: consider checking equality based only on the set of procedures
+        self.module == other.module && self.procedures == other.procedures
+    }
+}
 
-    pub fn serialize<S>(module: &super::ModuleAst, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: serde::Serializer,
-    {
-        let bytes = module.to_bytes(AstSerdeOptions {
-            serialize_imports: true,
-        });
+impl Eq for AccountCode {}
 
-        serializer.serialize_bytes(&bytes)
+// SERIALIZATION
+// ================================================================================================
+
+impl Serializable for AccountCode {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        self.module.write_into(target, MODULE_SERDE_OPTIONS);
+        self.module.write_source_locations(target);
+        // since the number of procedures is guaranteed to be between 1 and 256, we can store the
+        // number as a single byte - but we do have to subtract 1 to store 256 as 255.
+        target.write_u8(self.procedures.len() as u8 - 1);
```

Comment:
This should be:

```suggestion
        target.write_u8((self.procedures.len() - 1) as u8);
```

We want to first make sure the value fits in the `u8` range, and then do the casting.

Here is a test program:

```rust
use std::str::FromStr;
fn main() {
    let v = u32::from_str(&std::env::args().skip(1).next().unwrap()).unwrap();
    let v2 = v as u8 - 1;
    println!("{v2}");
}
```

The above panics when the CLI argument is `256`, because the `as u8` casts the value to `0`, and then we subtract one from that, since `u8` is unsigned it panics.

Reply:
Good catch! Fixed!